router/core/server.go

```diff
@@ -0,0 +1,684 @@
+package core
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"github.com/dgraph-io/ristretto"
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/chi/v5/middleware"
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+	"github.com/wundergraph/cosmo/router/gen/proto/wg/cosmo/common"
+	nodev1 "github.com/wundergraph/cosmo/router/gen/proto/wg/cosmo/node/v1"
+	"github.com/wundergraph/cosmo/router/internal/requestlogger"
+	"github.com/wundergraph/cosmo/router/internal/retrytransport"
+	"github.com/wundergraph/cosmo/router/pkg/config"
+	"github.com/wundergraph/cosmo/router/pkg/health"
+	rmetric "github.com/wundergraph/cosmo/router/pkg/metric"
+	"github.com/wundergraph/cosmo/router/pkg/otel"
+	"github.com/wundergraph/cosmo/router/pkg/pubsub"
+	"github.com/wundergraph/cosmo/router/pkg/pubsub/kafka"
+	pubsubNats "github.com/wundergraph/cosmo/router/pkg/pubsub/nats"
+	rtrace "github.com/wundergraph/cosmo/router/pkg/trace"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/engine/datasource/pubsub_datasource"
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+	"go.opentelemetry.io/otel/attribute"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	oteltrace "go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+const (
+	featureFlagHeader = "X-Feature-Flag"
+	featureFlagCookie = "feature_flag"
+)
+
+type (
+	// Server is the public interface of the server.
+	Server interface {
+		HttpServer() *http.Server
+		HealthChecks() health.Checker
+		BaseURL() string
+	}
+
+	EnginePubSubProviders struct {
+		nats  map[string]pubsub_datasource.NatsPubSub
+		kafka map[string]pubsub_datasource.KafkaPubSub
+	}
+
+	// server is the swappable implementation of the Router which is an HTTP server with middlewares.
+	// Everytime a schema is updated, the old server is shutdown and a new server is created.
+	// For feature flags, a separate mux is created and dynamically switched based on the feature flag header or cookie.
+	// All server fields are shared between all feature muxes.
+	server struct {
+		Config
+		httpServer              *http.Server
+		healthChecks            health.Checker
+		pubSubProviders         *EnginePubSubProviders
+		websocketStats          WebSocketsStatistics
+		playgroundHandler       func(http.Handler) http.Handler
+		publicKey               *ecdsa.PublicKey
+		executionTransport      *http.Transport
+		baseOtelAttributes      []attribute.KeyValue
+		runtimeMetrics          *rmetric.RuntimeMetrics
+		metricStore             rmetric.Store
+		baseRouterConfigVersion string
+	}
+)
+
+func (s *server) buildMultiGraphHandler(ctx context.Context, baseMux *chi.Mux, featureFlagConfigs map[string]*nodev1.FeatureFlagRouterExecutionConfig) (http.HandlerFunc, error) {
+
+	featureFlagToMux := make(map[string]*chi.Mux, len(featureFlagConfigs))
```

Comment:
can this func be shortcut with a simpler handler in case there are no FF configs or there is only one FF? In such a case, no map is needed and we don't need FF detection, right?

Reply:
Done